### PR TITLE
Add Wikibase tarball contents ADR

### PR DESCRIPTION
**Pull request by toban**
https://phabricator.wikimedia.org/T267754

### CODE REVIEW

File: docs/adr/0003-tarball-contents.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  - `.gitreview`
+  - `.git` folder
+  - `.github` folder
+  - `.vscode` folder
+  - `.phan` folder
+  - etc.
+
+## Consequences
+
+- Do not include release notes in the tarball.
```
Comment (mariushoch) [inline docs/adr/0003-tarball-contents.md:66]: This contradicts line 46… left over form an earlier state?

File: docs/adr/0003-tarball-contents.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+Seeing that there is an already defined process for release notes by the mediawiki community it feels natural to use existing tools and adopting this process.
+
+The `vendor` folder will not be included in the tarball as composer dependencies should be merged by the [composer merge plugin](https://github.com/wikimedia/composer-merge-plugin).  
+
+The tarball package *WILL* include a part from the source-code the following additions.
+
+- Release notes (RELEASE-NOTES-N.NN) (example RELEASE-NOTES-1.36)
+- Historic changes (HISTORY) (Previous release notes in one file)
+- git submodule dependencies
```
Comment (mariushoch) [inline docs/adr/0003-tarball-contents.md:49]: What does that mean? The submodules or their dependencies (however that might look)?

File: docs/adr/0003-tarball-contents.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+---
+
+MediaWiki sticks out from the crowd by including most of their release specific documentation in each of the tarball releases while all other projects in the survey do not. At first glance this feels like a undesirable pattern to adopt as it's uncommon among other projects. In reality this points out a tested and proven methodology for documenting changes and shipping release notes by the mediawiki community. Release notes are added to the release branch and refined throughout the release process and then used as a template for the documentation on mediawiki.org.     
+## Decision
+
+Seeing that there is an already defined process for release notes by the mediawiki community it feels natural to use existing tools and adopting this process.
+
+The `vendor` folder will not be included in the tarball as composer dependencies should be merged by the [composer merge plugin](https://github.com/wikimedia/composer-merge-plugin).  
+
+The tarball package *WILL* include a part from the source-code the following additions.
```
Comment (mariushoch) [inline docs/adr/0003-tarball-contents.md:45]: s/a part/apart/

**Review by mariushoch — CHANGES_REQUESTED** [review]

File: docs/adr/0003-tarball-contents.md (hunk 2 above)
Comment (toban) [reply docs/adr/0003-tarball-contents.md]: I was referring to the submodules but did not consider their dependencies. 

I guess the dependencies of those would be either in vendor or node_modules which would be excluded?

**Review by mariushoch — APPROVED** [review]

**Comment by mariushoch** [comment]
Sounds good to me now.

File: docs/adr/0003-tarball-contents.md (hunk 2 above)
Comment (mariushoch) [reply docs/adr/0003-tarball-contents.md]: > I guess the dependencies of those would be either in vendor or node_modules which would be excluded?

I think so... and everything needed for a production setup (like minified bundle files) should be in git (that's also how we deploy this in WMF production after all).

File: docs/adr/0003-tarball-contents.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+8. Readme.html with installation & upgrade instructions and links to documentation for further reading.
+
+---
+
+MediaWiki sticks out from the crowd by including most of their release specific documentation in each of the tarball releases while all other projects in the survey do not. At first glance this feels like a undesirable pattern to adopt as it's uncommon among other projects. In reality this points out a tested and proven methodology for documenting changes and shipping release notes by the mediawiki community. Release notes are added to the release branch and refined throughout the release process and then used as a template for the documentation on mediawiki.org.     
+## Decision
+
+Seeing that there is an already defined process for release notes by the mediawiki community it feels natural to use existing tools and adopting this process.
+
+The `vendor` folder will not be included in the tarball as composer dependencies should be merged by the [composer merge plugin](https://github.com/wikimedia/composer-merge-plugin).
```
Comment (tarrow) [inline docs/adr/0003-tarball-contents.md:43]: I'm still really unsure about this:
For example [the mediawiki installation requirements](https://www.mediawiki.org/wiki/Manual:Installation_requirements) don't include composer and [Installing mediawiki for developers](https://www.mediawiki.org/wiki/Manual:Installing_MediaWiki#Git_(for_developers)) seems to imply composer shouldn't need to be run when "just normally" installing

Comment (tarrow) [reply docs/adr/0003-tarball-contents.md:43]: [Wikibase install instructions](https://www.mediawiki.org/wiki/Wikibase/Installation) however heavily reference composer

Comment (mariushoch) [reply docs/adr/0003-tarball-contents.md:43]: From https://www.mediawiki.org/wiki/Manual:External_libraries:
> The extension distributor automatically packages composer dependencies, so tarball users won't have to do it manually.

`ExtensionProcessor::getExtraAutoloaderPaths` also seems to pick up these vendor folders… so I guess we might want to revise this? But I'm not sure how well this copes with conflicting dependency versions (I guess not well/ at all).

Comment (mariushoch) [reply docs/adr/0003-tarball-contents.md:43]: From https://www.mediawiki.org/wiki/Manual:External_libraries:
> The extension distributor automatically packages composer dependencies, so tarball users won't have to do it manually.

`ExtensionProcessor::getExtraAutoloaderPaths` also seems to pick up these vendor folders… so I guess we might want to revise this? But I'm not sure how well this copes with conflicting dependency versions (I guess not well/ at all).

Comment (toban) [reply docs/adr/0003-tarball-contents.md:43]: I poked the release team about their reasoning for including the `vendor` folder in and the arguments were the following

- Not everyone can run composer (security, shared hosting, shell access etc)
- Including it we can support both people that has access to composer and people that don't (It's easier to rm -rf vendor and the composer.lock (if you don't need them) than it is to add them later)

original task: https://phabricator.wikimedia.org/T70940

I can agree this makes sense to me. But we would probably need to update the documentation as you say @tarrow.

Comment (toban) [reply docs/adr/0003-tarball-contents.md]: I did a poor job at investigating the mediawiki tarballs - they indeed include the vendor.

**Pull request merged**